SCClassLibrary/Common/Collections/ArrayedCollection.sc

```diff
@@ -321,14 +321,21 @@ ArrayedCollection : SequenceableCollection {
 	}
 
 	// concepts borrowed from J programming language
-	rank {
-		// rank is the number of dimensions in a multidimensional array.
-		// see also Object-rank
-		// this assumes every element has the same rank
+	isRectangular {
+	    // all nested arrays are the same size and elements are at the same depth.
+	    _ArrayIsRectangular
+		^this.primitiveFailed
+	}
+	rank {|test=false|
+	    if(test and: this.isRectangular.not){
```

Comment:
You need braces `if(test and: { this.isRectangular.not }){` for this to optimize.

Reply:
The issues is that it has to visit every element O(n). 

```supercollider

// 1
a = [];
bench { a.isRectangular }; // time to run: 1.432999994222e-06 seconds.

// 2
a = (0..1000);
bench { a.isRectangular }; // time to run: 5.7009999920865e-06 seconds.

// 3
a = (0..(1000*1000));
bench { a.isRectangular }; // time to run: 0.0024116329999941 seconds.

// 4
a = 0!1000!1000;
bench { a.isRectangular }; // time to run: 0.0026010189999965 seconds.

// 5
a = 0!1000!1000!1000; // time to run: 2.506840667 seconds.
bench { a.isRectangular };
```
You can see that 3 and 4 are close in performance as they have the same number of elements.

Reply:
I've remove the test